content/blog/2023-12-11-jitnic/index.md

```diff
@@ -0,0 +1,129 @@
+    +++
+    title = "JITNIC"
+    [extra]
+    bio = """
+      Zak Kent is a senior interested in programming languages and verification.
+      Benny Rubin is a junior interested in networking and systems.
+    """
+    [[extra.authors]]
+    name = "Zak Kent"
+    [[extra.authors]]
+    name = "Benny Rubin"
+    +++
+
+## Summary
+
+For this project, we implemented an eBPF JIT compiler to share eXpress Data Path (XDP) network programs between a host CPU and a smartNIC. The fundamental idea behind this design is to use runtime information to compile and run straight-line traces of eBPF code on a smartNIC. Straight line traces of code enable both compiler level and hardware level optimizations. The key challenge to this approach is to decide if a given trace can correctly run on a packet and to dynamically recompile a new trace if not. We develop the idea of a “path condition” - a predicate over fields of a packet. Each trace has a corresponding unique minimal path condition that is sufficient to determine the set of incoming packets that can be run on it. Finally, we evaluate our approach by looking at worst case and best case trace compilation rates, where fewer recompilations is better for performance. We find in practice that XDP programs require few traces installed on a NIC to cover the majority of incoming packets for realistic workloads.
```

Comment:
> We develop the idea of a “path condition”

I might slightly rephrase this; it makes it sounds a bit like you invented the idea of "path conditions", which are of course an integral part of symbolic execution.